File: app/src/main/java/com/shevaalex/android/rickmortydatabase/ui/episode/EpisodeAdapter.java
package com.shevaalex.android.rickmortydatabase.ui.episode;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.shevaalex.android.rickmortydatabase.R;
import com.shevaalex.android.rickmortydatabase.databinding.ItemEpisodeBinding;
import com.shevaalex.android.rickmortydatabase.models.episode.EpisodeModel;

import me.zhanghai.android.fastscroll.PopupTextProvider;

public class EpisodeAdapter
        extends PagedListAdapter<EpisodeModel, EpisodeAdapter.EpisodeViewHolder>
        implements PopupTextProvider {
    private final OnEpisodeClickListener clickListener;
    private final Context context;

    EpisodeAdapter(Context context, OnEpisodeClickListener clickListener) {
        super(DIFF_CALLBACK);
        this.clickListener = clickListener;
        this.context = context;
    }

    private static final DiffUtil.ItemCallback<EpisodeModel> DIFF_CALLBACK
            = new DiffUtil.ItemCallback<EpisodeModel>() {
        @Override
        public boolean areItemsTheSame(@NonNull EpisodeModel oldItem, @NonNull EpisodeModel newItem) {
            return oldItem.getId() == newItem.getId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull EpisodeModel oldItem, @NonNull EpisodeModel newItem) {
            return newItem.equals(oldItem);
        }
    };

    @NonNull
    @Override
    public EpisodeViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        //instantiate view binding class and pass it to ViewHolder
        ItemEpisodeBinding binding = ItemEpisodeBinding
                .inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new EpisodeViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull EpisodeViewHolder holder, int position) {
        EpisodeModel currentEpisode = getItem(position);
        if (currentEpisode != null) {
            holder.binding.episodeNameValue
                    .setText(String.format(context.getResources().getString(R.string.episode_name_placeholder), currentEpisode.getName()));
            if (holder.binding.episodeAirDateValue != null) {
                holder.binding.episodeAirDateValue.setText(currentEpisode.getAirDate());
            }
            holder.binding.episodeCodeValue.setText(currentEpisode.getCode());
        }
    }

    @NonNull
    @Override
    public String getPopupText(int position) {
        EpisodeModel currentEpisode = getItem(position);
        if (currentEpisode != null) {
            return "s" + currentEpisode.getCode().substring(2, 3)
                    + "e" + currentEpisode.getCode().substring(4);
        } else {
            return "";
        }
    }

    class EpisodeViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private final ItemEpisodeBinding binding;

        EpisodeViewHolder(ItemEpisodeBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            if (getAbsoluteAdapterPosition() != RecyclerView.NO_POSITION) {
                clickListener.onEpisodeClick(getAbsoluteAdapterPosition(), v);
            }
        }
    }

    public interface OnEpisodeClickListener {
        void onEpisodeClick(int position, View v);
    }
}

File: app/src/main/java/com/shevaalex/android/rickmortydatabase/ui/episode/list/EpisodeAdapter.kt
package com.shevaalex.android.rickmortydatabase.ui.episode.list

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import com.google.android.material.card.MaterialCardView
import com.shevaalex.android.rickmortydatabase.R
import com.shevaalex.android.rickmortydatabase.databinding.ItemEpisodeBinding
import com.shevaalex.android.rickmortydatabase.models.episode.EpisodeEntity
import com.shevaalex.android.rickmortydatabase.utils.Constants.Companion.TRANSITION_EPISODE

class EpisodeAdapter(
        private val placeHolderString: String,
        private val episodeListener: EpisodeListener
) : PagedListAdapter<EpisodeEntity, EpisodeAdapter.EpisodeViewHolder>(DIFF_CALLBACK) {

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<EpisodeEntity>() {
            override fun areItemsTheSame(oldItem: EpisodeEntity, newItem: EpisodeEntity): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areContentsTheSame(oldItem: EpisodeEntity, newItem: EpisodeEntity): Boolean {
                return newItem == oldItem
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): EpisodeViewHolder {
        val itemBind = ItemEpisodeBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
        )
        return EpisodeViewHolder(itemBind)
    }

    override fun onBindViewHolder(holder: EpisodeViewHolder, position: Int) {
        val episode = getItem(position)
        episode?.let {
            holder.bind(episode, episodeListener, placeHolderString)
        }
    }

    class EpisodeViewHolder(
            private val itemBind: ItemEpisodeBinding
    ) : RecyclerView.ViewHolder(itemBind.root) {

        fun bind(episode: EpisodeEntity, episodeListener: EpisodeListener, placeHolderString: String) {
            val context = itemBind.root.context
            itemBind.root.setOnClickListener {
                episodeListener.onEpisodeClick(episode, itemBind.episodeItem)
            }
            itemBind.episodeItem.transitionName = TRANSITION_EPISODE.plus(episode.id)
            Glide.with(context)
                    .load(episode.imageUrl)
                    .override(284, 200)
                    .apply(RequestOptions()
                            .placeholder(R.drawable.episode_placeholder)
                            .diskCacheStrategy(DiskCacheStrategy.DATA)
                    )
                    .into(itemBind.episodeImage)
            itemBind.episodeNameValue.text = placeHolderString.format(episode.name)
            itemBind.episodeAirDateValue?.let {
                it.text = episode.airDate
            }
            itemBind.episodeCodeValue.text = episode.code
        }

    }

    interface EpisodeListener {
        fun onEpisodeClick(episode: EpisodeEntity, episodeCard: MaterialCardView)
    }

}